app/services/segment_delete_service.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+class SegmentDeleteService
+  SEGMENT_URI = URI('https://gdpr.segment.com/graphql')
+
+  class SegmentDeleteServiceError < StandardError
+    def initialize(message:, response:)
+      super(message)
+      @response = {body: response.body, code: response.code}
+    end
+
+    attr_reader :response
+
+    def error_type
+      self.class.name.gsub(/\ASegmentDeleteService::Segment|Error\z/, '').underscore.to_sym
```

Comment:
I think this is for https://github.com/3scale/porta/blob/0be22b3adf10b8e36bb8895e38e7a105afc8c866/app/services/segment_delete_service.rb#L48
Right?

Maybe we could keep it bit simpler:

```ruby
class SegmentDeleteServiceError < StandardError
  def error_type
    raise NotImplementedError
  end
end

class SegmentAuthenticationError < SegmentDeleteServiceError
  def error_type
    :authentication
  end
end

class SegmentUserDeletionError < SegmentDeleteServiceError
  def error_type
    :user_deletion
  end
end
```

So superclass does not have to know about subclases.

Reply:
Ok, it makes sense, but I prefer this way then:

```ruby
class SegmentDeleteServiceError < StandardError
  def error_type
    :delete_service
  end
end

class SegmentAuthenticationError < SegmentDeleteServiceError
  def error_type
    :authentication
  end
end

class SegmentUserDeletionError < SegmentDeleteServiceError
  def error_type
    :user_deletion
  end
end
```

because I don't want the father to be an abstract class, I want it to be usable.